cartoframes/data/dataset.py

```diff
@@ -1,282 +1,106 @@
-
-        Example:
-
-            Create a Dataset from a pandas Dataframe and then map the data.
-
-            .. code::
-
-                from cartoframes.data import Dataset
-                from cartoframes.viz import Map, Layer
-                import pandas as pd
-
-                df = pd.DataFrame({'lat': [0, 10, 20], 'lng': [20, 10, 0]})
-
-                d = Dataset.from_dataframe(df)
-
-                Map(Layer(d))
-
-        """
-        dataset = cls(df=df, state=cls.STATE_LOCAL)
-        _save_index_as_column(dataset._df)
-        return dataset
-
-    @classmethod
-    def from_geodataframe(cls, gdf):
-        """Create a Dataset from a local GeoPandas GeoDataFrame.
-
-        Args:
-          gdf (geopandas.GeoDataFrame): GeoPandas GeoDataFrame
-
-        Example:
-
-          GeoDataFrame example code taken from `GeoPandas documentation
-          <http://geopandas.org/gallery/create_geopandas_from_pandas.html#creating-a-geodataframe-from-a-dataframe-with-coordinates>`__.
-
-        .. code::
-
-            from cartoframes.data import Dataset
-            from cartoframes.viz import Map, Layer
-            import pandas as pd
-            import geopandas as gpd
-
-            df = pd.DataFrame(
-                {'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
-                 'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
-                 'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
-                 'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]})
-            gdf = gpd.GeoDataFrame(
-                df,
-                geometry=gpd.points_from_xy(df.Longitude, df.Latitude)
-            )
-
-            d = Dataset.from_geodataframe(gdf)
-
-            Map(Layer(d))
-
-        """
-        dataset = cls(df=gdf, state=cls.STATE_LOCAL)
-        _save_index_as_column(dataset._df)
-        return dataset
-
-    @classmethod
-    def from_geojson(cls, geojson):
-        """Create a Dataset from a GeoJSON file (hosted or local).
-
-        Args:
-          gdf (geopandas.GeoDataFrame): GeoPandas GeoDataFrame
-
-        Example:
+    GEOM_TYPE_POINT = GEOM_TYPE_POINT
+    GEOM_TYPE_LINE = GEOM_TYPE_LINE
+    GEOM_TYPE_POLYGON = GEOM_TYPE_POLYGON
+
+    def __init__(self, data, credentials=None, schema=None):
+        self._strategy = self._init_strategy(data, credentials, schema)
+        self._is_saved_in_carto = self._init_saved_in_carto()
+
+    def _init_strategy(self, data, credentials=None, schema=None):
+        credentials = credentials or _get_default_credentials()
+
+        if isinstance(data, pd.DataFrame):
+            return self._getDataFrameDataset(data)
+        elif isinstance(data, (list, dict)):
+            return self._getDataFrameDataset(load_geojson(data))
+        elif isinstance(data, str):
+            if is_sql_query(data):
+                if not credentials:
+                    raise ValueError('QueryDataset needs a Credentials object')
```

Comment:
These raises should be done inside the `self._getWhateverDataset(data, credentials)` method

Reply:
In the `self._getWhateverDataset` I am creating a new instance (making the code testable). In some moment in the past, I moved the raises to there, but  `self._getWhateverDataset`  would be doing 2 things instead of one.

- In `_init_strategy` is where we have the logic the choose the instance. Probably you are right and it is not the place for validations.

We could move it into the constructor instances:

```

class QueryDataset(BaseDataset):
    def __init__(self, data, credentials=None, schema=None):
        if not credentials:
            raise ValueError('QueryDataset needs a Credentials object')

        super(QueryDataset, self).__init__(credentials)

        self._query = data
```

How do you see it?

Reply:
Seeing my proposal again, it is weird having `credentials=None` and raising an exception if `not credentials`

Reply:
You need to take into account the `default_credentials`

Reply:
^ `credentials = credentials or _get_default_credentials()`

Reply:
I have changed both Query and Table datasets ensuring a credentials object. For instance:

```
class QueryDataset(BaseDataset):
    def __init__(self, data, credentials, schema=None):
```

And Python will do the validation. This is the good aproach and it is what we had before. I had been trying to avoid a problem revolved by the language.